src/ut_control_plane.c

```diff
@@ -0,0 +1,397 @@
+    return NULL;
+}
+
+static int callback_echo(struct lws *wsi, enum lws_callback_reasons reason, void *user, void *in, size_t len)
+{
+    switch (reason)
+    {
+    case LWS_CALLBACK_ESTABLISHED:
+        UT_CONTROL_PLANE_DEBUG("Client connected\n");
+        break;
+
+    case LWS_CALLBACK_RECEIVE:
+        UT_CONTROL_PLANE_DEBUG("LWS_CALLBACK_RECEIVE\n");
+        cp_message_t msg;
+        msg.message = malloc((int)len + 1);
+        if(msg.message == NULL)
+        {
+            UT_CONTROL_PLANE_ERROR("Malloc failed\n");
+            break;
+        }
+        msg.size = (int)len;
+        msg.status = DATA_RECIEVED;
+        memset(msg.message, 0, (int)len + 1);
+        strncpy(msg.message, (const char*)in, len);
+        msg.message[len] = '\0';
+        UT_CONTROL_PLANE_DEBUG("Received message:\n %s\n", msg.message);
+        enqueue_message(&msg);
+        // Echo back received message
+        //lws_write(wsi, in, len, LWS_WRITE_TEXT);
+        break;
+
+    default:
+        break;
+    }
+
+    return 0;
+}
+
+static struct lws_protocols protocols[] = {
+    {
+        "echo-protocol",    // protocol name
+        callback_echo,      // callback function
+        0,                  // per-session data size
+        0,                  // rx buffer size
+        0,                  // id
+        NULL,               // reserved
+        0                   // tx packet size
+    },
+    { NULL, NULL, 0, 0 }    // end of list
+};
+
+ut_controlPlane_instance_t *UT_ControlPlane_Init( uint32_t monitorPort )
+{
+    ut_cp_instance_internal_t *pInstance = calloc(1, sizeof(ut_cp_instance_internal_t));
+
+    if ( pInstance == NULL )
+    {
+        //assert( pInstance != NULL );
+        UT_CONTROL_PLANE_ERROR("Malloc was not able to provide memory\n");
+        return NULL;
+    }
+
+    pInstance->info.port = monitorPort;
+    pInstance->info.iface = NULL;
+    pInstance->info.protocols = protocols;
+    pInstance->info.user = pInstance;
+
+    pInstance->context = lws_create_context(&pInstance->info);
+    if (pInstance->context == NULL)
+    {
+        UT_CONTROL_PLANE_ERROR("Error creating libwebsockets context\n");
+        free( pInstance );
+        return NULL;
+    }
+
+    pInstance->exit_request = false;
+    pInstance->callback_entry_index = 0;
+    pInstance->magic = UT_CP_MAGIC;
+
+    return (ut_controlPlane_instance_t *)pInstance;
+}
+
+void UT_ControlPlane_Exit( ut_controlPlane_instance_t *pInstance )
+{
+    ut_cp_instance_internal_t *pInternal = validateCPInstance(pInstance);
+
+    if (pInternal == NULL)
+    {
+        return;
+    }
+
+    if (pInternal->context == NULL)
+    {
+        UT_CONTROL_PLANE_ERROR("libwebsockets context was not created\n");
+        return;
+    }
+
+    if (pInternal->state_machine_thread_handle )
+    {
+        cp_message_t msg;
+        msg.status = EXIT_REQUESTED;
+        UT_CONTROL_PLANE_ERROR("EXIT_REQUESTED message from [%s] \n", __func__);
```

Comment:
For some strange reason printf() is helping here to capture the CTRL-C from sig handler. Otherwise, it catches the signal on second attempt and then we get message that pthread_join fails. 
```
2024-06-13-10:23:20, LOG   , ut_test_control_plane.c,   127 : Please Run the command `test1DataSend.sh` and press return;'^C
^CFailed to join state_machine_thread_handle(1st) from instance = 0x560aa597ad80
[2024/06/13 10:23:33:0796] N: __lws_lc_untag:  -- [wsi|0|pipe] (1) 12.840s
[2024/06/13 10:23:33:0797] N: __lws_lc_untag:  -- [vh|1|default||8080] (1) 12.840s
[2024/06/13 10:23:33:0797] N: __lws_lc_untag:  -- [wsi|1|listen|default||8080] (0) 12.840s
[2024/06/13 10:23:33:0797] N: __lws_lc_untag:  -- [vh|0|netlink] (0) 12.840s
^C
Invalid Handle - magic failure^C
Invalid Handle - magic failureq
q
^C
Invalid Handle - magic failure^C
Invalid Handle - magic failure^C
Invalid Handle - magic failure^C
Invalid Handle - magic failure^Z
[1]+  Stopped                 ./tests/bin/ut_control_test.sh
```
Couldn't find any reasoning but this helps.

Reply:
I moved logging to test file and it works fine.